File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = head;
        ListNode temp2 = head;
        int count = 0;

        ListNode counter = head;
        while (counter != null) {
            counter = counter.next;
            count++;
        }

        if (n == count) {
            return head.next;
        }

        int cur = 0;
        while (cur < count - n - 1) {  
            temp2 = temp2.next;
            cur++;
        }

        temp2.next = temp2.next.next;

        return temp;
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            # Find the kth node
            kth = groupPrev
            for i in range(k):
                kth = kth.next
                if not kth:
                    return dummy.next

            groupNext = kth.next

            groupStart = groupPrev.next

            kth.next = None

            reversedHead = self.reverseLinkedList(groupStart)

            groupPrev.next = reversedHead
            groupStart.next = groupNext

            groupPrev = groupStart


    def reverseLinkedList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        cur = head

        while cur:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp

        return prev